# **Cafe Sales - Dirty Data for Cleaning Training**

## Overview

The Dirty Cafe Sales dataset contains 10,000 rows of synthetic data representing sales transactions in a cafe. This dataset is intentionally "dirty," with missing values, inconsistent data, and errors introduced to provide a realistic scenario for data cleaning and exploratory data analysis (EDA). It can be used to practice cleaning techniques, data wrangling, and feature engineering.

### File Information

- File Name: `dirty_cafe_sales.csv`
- Number of Rows: `10,000`
- Number of Columns: `8`

### Columns Description

Column Name  |Description	 |Example Values  |
-----|-----|-----|
Transaction ID |A unique identifier for each transaction. Always present and unique. |`TXN_1234567` |
Item |The name of the item purchased. May contain missing or invalid values (e.g., "ERROR"). |`Coffee, Sandwich` |
Quantity |The quantity of the item purchased. May contain missing or invalid values. |`1, 3, UNKNOWN` |
Price Per Unit |The price of a single unit of the item. May contain missing or invalid values. |`2.00, 4.00` |
Total Spent |The total amount spent on the transaction. Calculated as Quantity * Price Per Unit. |`8.00, 12.00` |
Payment Method |The method of payment used. May contain missing or invalid values (e.g., None, "UNKNOWN"). |`Cash, Credit Card`|
Location |The location where the transaction occurred. May contain missing or invalid values |`In-store, Takeaway` |
Transaction Date |The date of the transaction. May contain missing or incorrect values. |`2023-01-01`|

### Import Packages

In [1]:
# Import packages
### YOUR CODE HERE ### 
# For data manipulation
import numpy as np
import pandas as pd

# For data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# For displaying all of the columns in dataframes
pd.set_option('display.max_columns', None)

## Load Dataset 

In [2]:
# Load dataset into a dataframe
df0 = pd.read_csv("dirty_cafe_sales.csv")

In [3]:
df0.head(10)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0    TXN_1961373    Coffee        2            2.0         4.0   
1    TXN_4977031      Cake        4            3.0        12.0   
2    TXN_4271903    Cookie        4            1.0       ERROR   
3    TXN_7034554     Salad        2            5.0        10.0   
4    TXN_3160411    Coffee        2            2.0         4.0   
5    TXN_2602893  Smoothie        5            4.0        20.0   
6    TXN_4433211   UNKNOWN        3            3.0         9.0   
7    TXN_6699534  Sandwich        4            4.0        16.0   
8    TXN_4717867       NaN        5            3.0        15.0   
9    TXN_2064365  Sandwich        5            4.0        20.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

## Data Exploration

### Basic information about the data 

In [4]:
# Gather descriptive statistics about the data
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


### Descriptitve statistics about the data

In [5]:
# Gather descriptive statistics about the data
### YOUR CODE HERE ###
df0.describe()

Transaction ID   Item Quantity Price Per Unit Total Spent  \
count           10000   9667     9862           9821        9827   
unique          10000     10        7              8          19   
top       TXN_1961373  Juice        5            3.0         6.0   
freq                1   1171     2013           2429         979   

        Payment Method  Location Transaction Date  
count             7421      6735             9841  
unique               5         4              367  
top     Digital Wallet  Takeaway          UNKNOWN  
freq              2291      3022              159

### Rename columns

In [6]:
# Display all column names
df0. columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='object')

In [7]:
# Rename columns as needed
df0 = df0.rename(
    columns = {'Transaction ID':'transaction_id',
               'Item':'item', 'Quantity':'quantity',
               'Price Per Unit':'price_per_unit',
               'Total Spent':'total_spent',
               'Payment Method':'payment_method',
               'Location':'location',
               'Transaction Date':'transaction_date'
              }
)

# Displaying the new name columns
df0.columns

Index(['transaction_id', 'item', 'quantity', 'price_per_unit', 'total_spent',
       'payment_method', 'location', 'transaction_date'],
      dtype='object')

### Create a copy

In [8]:
#Creating a copy from the dataframe
df = df0.copy()

### Replacing the UNKNOWN and ERROR values

In [9]:
cat_col = {'item', 'quantity', 'price_per_unit', 'total_spent',
       'payment_method', 'location'}
#Determining the unique values for most of the columns
for col in cat_col:
    unique = df[col].unique()
    print(f"{col} Values: {unique}\n")

total_spent Values: ['4.0' '12.0' 'ERROR' '10.0' '20.0' '9.0' '16.0' '15.0' '25.0' '8.0' '5.0'
 '3.0' '6.0' nan 'UNKNOWN' '2.0' '1.0' '7.5' '4.5' '1.5']

price_per_unit Values: ['2.0' '3.0' '1.0' '5.0' '4.0' '1.5' nan 'ERROR' 'UNKNOWN']

location Values: ['Takeaway' 'In-store' 'UNKNOWN' nan 'ERROR']

quantity Values: ['2' '4' '5' '3' '1' 'ERROR' 'UNKNOWN' nan]

item Values: ['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'UNKNOWN' 'Sandwich' nan
 'ERROR' 'Juice' 'Tea']

payment_method Values: ['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR' nan]



In [10]:
#Replacing with NaN values
df.replace(['UNKNOWN', 'ERROR'], np.nan, inplace = True)

#Repeating the process above to check that the values were correctly replaced.
cat_col = {'item', 'quantity', 'price_per_unit', 'total_spent',
       'payment_method', 'location'}

for col in cat_col:
    unique = df[col].unique()
    print(f"{col} Values: {unique}\n")

total_spent Values: ['4.0' '12.0' nan '10.0' '20.0' '9.0' '16.0' '15.0' '25.0' '8.0' '5.0'
 '3.0' '6.0' '2.0' '1.0' '7.5' '4.5' '1.5']

price_per_unit Values: ['2.0' '3.0' '1.0' '5.0' '4.0' '1.5' nan]

location Values: ['Takeaway' 'In-store' nan]

quantity Values: ['2' '4' '5' '3' '1' nan]

item Values: ['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' nan 'Sandwich' 'Juice' 'Tea']

payment_method Values: ['Credit Card' 'Cash' nan 'Digital Wallet']



In [11]:
print("NaN/Null values per column", df.isnull().sum(), "\n")
print("Total rows with NaN/Null values", df.isnull().any(axis=1).sum())

NaN/Null values per column transaction_id         0
item                 969
quantity             479
price_per_unit       533
total_spent          502
payment_method      3178
location            3961
transaction_date     460
dtype: int64 

Total rows with NaN/Null values 6911


## Data Cleaning

### Changing the Data type

In [12]:
int_col = {'quantity', 'price_per_unit', 'total_spent'}

#Transforming the data type from these three columns into numeric
for col in int_col:
    df[col] = pd.to_numeric(df[col], errors = 'coerce')

#Checking the data type
df.dtypes

transaction_id       object
item                 object
quantity            float64
price_per_unit      float64
total_spent         float64
payment_method       object
location             object
transaction_date     object
dtype: object

### Fixing Price Per Unit List

Determining the price list per item.

In [13]:
df[['item', 'price_per_unit']].dropna().drop_duplicates().sort_values(by='price_per_unit')

item  price_per_unit
2     Cookie             1.0
42       Tea             1.5
0     Coffee             2.0
1       Cake             3.0
17     Juice             3.0
7   Sandwich             4.0
5   Smoothie             4.0
3      Salad             5.0

Since there are `NaN` values in the columns `price_per_unit`, `quantity`, `total_spent`, I filled it by using the following formula:
$$t = {q}*{p}$$

Where:
- `t` = total_spent
- `q` = quantity
- `p` = price_per_unit

In [14]:
#Operations for the following columns.
df['price_per_unit'] = df['price_per_unit'].fillna(df['total_spent']/df['quantity'])
df['quantity'] = df['quantity'].fillna(df['total_spent']/df['price_per_unit'])
df['total_spent'] = df['total_spent'].fillna(df['quantity'] * df['price_per_unit'])

In [15]:
#Dictionarie to fix the prices for each item.
fixed_prices = {
    'Cookie':1.0,
    'Tea':1.5,
    'Coffee':2.0,
    'Cake':3.0,
    'Juice':3.0,
    'Sandwich':4.0,
    'Smoothie':4.0,
    'Salad':5.0
}

#Function to fix the prices for each item.
def correct_price(row):
    item = row['item']
    if item in fixed_prices:
        return fixed_prices[item]
    else:
        return row['price_per_unit']

df['price_per_unit'] = df.apply(correct_price, axis=1)

print("Total rows with NaN/Null values", df.isnull().any(axis=1).sum())

Total rows with NaN/Null values 6444


We reduced the `Nan`/Null values from 6911 to 6444. Although, we can reduced a little bit more by doing the other way around by fixing the items for each price that does not repeeat `price_per_unit`.

In [16]:
fixed_items = {
    1.0: 'Cookie',
    1.5: 'Tea',
    2.0: 'Coffee',
    5.0: 'Salad'
}

def correct_items(row):
    price = row['price_per_unit']
    if price in fixed_items:
        return fixed_items[price]
    else:
        return row['item']

df['item'] = df.apply(correct_items, axis=1)

print("Total rows with NaN/Null values", df.isnull().any(axis=1).sum())

Total rows with NaN/Null values 6243


### Convert Transaction Date data type to Datatime Format.

In [17]:
df['transaction_date'] = pd.to_datetime(df['transaction_date'], format='%Y-%m-%d', errors='coerce')
df['transaction_date']

0      2023-09-08
1      2023-05-16
2      2023-07-19
3      2023-04-27
4      2023-06-11
          ...    
9995   2023-08-30
9996   2023-06-02
9997   2023-03-02
9998   2023-12-02
9999   2023-11-07
Name: transaction_date, Length: 10000, dtype: datetime64[ns]

## Data Imputation

We run `info()` command to determine the Non-Null count for each column. It's important to remark that these steps is to fill these values with consistent steps to later data treatment.

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    10000 non-null  object        
 1   item              9520 non-null   object        
 2   quantity          9962 non-null   float64       
 3   price_per_unit    9994 non-null   float64       
 4   total_spent       9960 non-null   float64       
 5   payment_method    6822 non-null   object        
 6   location          6039 non-null   object        
 7   transaction_date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB


### Imputing missing Data

We fill the NaN values for the `quantity` and `price_per_unit` by using the mean function. On the other hand, the `total_spent` we just simply re used the formula shown above.

In [19]:
impute_f = ['quantity', 'price_per_unit']
df[impute_f] = df[impute_f].fillna(df[impute_f].mean())
df['total_spent'] = df['total_spent'].fillna(df['quantity'] * df['price_per_unit'])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    10000 non-null  object        
 1   item              9520 non-null   object        
 2   quantity          10000 non-null  float64       
 3   price_per_unit    10000 non-null  float64       
 4   total_spent       10000 non-null  float64       
 5   payment_method    6822 non-null   object        
 6   location          6039 non-null   object        
 7   transaction_date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB


To preserve the data size, I filled the NaN values of the categorical columns with "Unknown"

In [20]:
impute_cat = ['item', 'payment_method', 'location']
for x in impute_cat:
    df[x].fillna("Unknown", inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    10000 non-null  object        
 1   item              10000 non-null  object        
 2   quantity          10000 non-null  float64       
 3   price_per_unit    10000 non-null  float64       
 4   total_spent       10000 non-null  float64       
 5   payment_method    10000 non-null  object        
 6   location          10000 non-null  object        
 7   transaction_date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB


C:\Users\Usuario\AppData\Local\Temp\ipykernel_19608\3666663414.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[x].fillna("Unknown", inplace = True)


### Droping NaN Vallues

I dropped the rows with NaN values because there are only only 460 rows with these values

In [21]:
df.dropna(subset=['transaction_date'], inplace=True)

## Overview of Clean DataSet

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9540 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    9540 non-null   object        
 1   item              9540 non-null   object        
 2   quantity          9540 non-null   float64       
 3   price_per_unit    9540 non-null   float64       
 4   total_spent       9540 non-null   float64       
 5   payment_method    9540 non-null   object        
 6   location          9540 non-null   object        
 7   transaction_date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 670.8+ KB


In [23]:
df.describe()

quantity  price_per_unit  total_spent               transaction_date
count  9540.000000     9540.000000  9540.000000                           9540
mean      3.022004        2.948604     8.923689  2023-07-01 23:00:31.698113536
min       1.000000        1.000000     1.000000            2023-01-01 00:00:00
25%       2.000000        2.000000     4.000000            2023-04-01 00:00:00
50%       3.000000        3.000000     8.000000            2023-07-02 00:00:00
75%       4.000000        4.000000    12.000000            2023-10-02 00:00:00
max       5.000000        5.000000    25.000000            2023-12-31 00:00:00
std       1.417191        1.278714     6.001516                            NaN